daemon/logger/loggerutils/rotatefilewriter.go

```diff
@@ -90,20 +96,125 @@ func rotate(name string, maxFiles int) error {
 	if maxFiles < 2 {
 		return nil
 	}
-	for i := maxFiles - 1; i > 1; i-- {
-		toPath := name + "." + strconv.Itoa(i)
-		fromPath := name + "." + strconv.Itoa(i-1)
+
+	for i := maxFiles - 1; i > 2; i-- {
+		toPath := name + "." + strconv.Itoa(i) + ".gz"
+		fromPath := name + "." + strconv.Itoa(i-1) + ".gz"
 		if err := os.Rename(fromPath, toPath); err != nil && !os.IsNotExist(err) {
 			return err
 		}
 	}
 
+	if _, err := os.Stat(name + ".1"); err == nil && maxFiles > 2 {
+		if err := os.Rename(name+".1", name+".2"); err != nil {
+			return err
+		}
+
+		if err := compressJSONFile(name + ".2"); err != nil {
+			return err
+		}
+	}
+
+	// The "[name].1" that jast renamed from "[name]" is not compressed
+	// in order to prevent the log tracking tool from losing some historical
+	// log data when a new log file is created.
 	if err := os.Rename(name, name+".1"); err != nil && !os.IsNotExist(err) {
 		return err
 	}
 	return nil
 }
 
+func compressJSONFile(fileName string) (err error) {
+	file, err := os.Open(fileName)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		file.Close()
+		os.Remove(fileName)
+	}()
+
+	outFile, err := os.OpenFile(fileName+".gz", os.O_CREATE|os.O_RDWR, 0640)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		outFile.Close()
+		if err != nil {
+			os.Remove(fileName + ".gz")
+		}
+	}()
+
+	compressWriter := gzip.NewWriter(outFile)
+	defer compressWriter.Close()
+
+	// Extract the last log entry timestramp from log file
+	lastLine, err := readLastLine(file)
+	l := &jsonlog.JSONLog{}
+	err = json.Unmarshal(lastLine, l)
+	if err != nil {
+		return err
+	}
+	lastTime := l.Created.Format(jsonlog.RFC3339NanoFixed)
+
+	// Add the last log entry timestramp to the gzip header
+	extra := &RotateFileMetadata{}
+	extra.LastTime = lastTime
+	compressWriter.Header.Extra, err = json.Marshal(extra)
+	if err != nil {
+		return err
+	}
+
+	file.Seek(0, os.SEEK_SET)
+	_, err = io.Copy(compressWriter, file)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func readLastLine(file *os.File) ([]byte, error) {
```

Comment:
>Maybe we can keep the last timestamp in memory somehow... This just being a generic writer I'm not sure right now the best approach.
>
>:( This may be getting too much in the way of the log writer in the name of optimizing reads... will have to think on this. Do you have any ideas?

@cpuguy83  I originally thought we should not compress the `container-id.log.1` log file so that other log tracking tools do not lose any history log when a new log file is created. In this case, we have to read the `last timestamp` from the file, because the data in the memory will be lost when the `dockerd ` is restarted.

However, if `jsonfilelog` is only used for the `docker log`, then lasttime is stored in memory is the better solution. :+1: